New survey page based on Next.js v13 using the app router and server actions

henrycatalinismith opened the pull request:
Combining all our work from https://github.com/zetkin/app.zetkin.org/pull/1636, https://github.com/zetkin/app.zetkin.org/pull/1719 and https://github.com/zetkin/app.zetkin.org/issues/1738. Targeting the branch for https://github.com/zetkin/app.zetkin.org/pull/1719 as the code changes here depend on merging the Next v13 upgrade to `main`. Once https://github.com/zetkin/app.zetkin.org/pull/1719 ships we can update the base branch of this PR to `main`.

henrycatalinismith commented:
Nice. The playwright failures here make sense. It's the result of combining the new playwright tests we wrote at code camp with the server action version of the survey form.

henrycatalinismith commented:
First little burst of activity since the hackathon produced [5459f24](https://github.com/zetkin/app.zetkin.org/pull/1756/commits/5459f24c85ec8b8178007c995da28029c780963f). The context here is I'm trying to get the survey page playwright tests green, and the executive summary is that this commit gets 2/5 of them green and then skips the remaining three. More details below.

1. ✅ `submits responses`
This one was failing because it was trying to sign as the logged-in user. That's a detail we agreed to cut from the initial scope of the app router migration of this feature, so the reason it doesn't work is simply that it's not implemented. Switching to anonymous signing is fine here as a compromise because the test isn't about the signature.

2. ✅ `submits email signature`
This one was failing because I had moved the survey data fetching to [`<SurveyForm />`](https://github.com/zetkin/app.zetkin.org/blob/issue-1620/next-v13-survey-page-with-app-router-and-server-action/src/features/surveys/components/surveyForm/SurveyForm.tsx), which gets its API client instance for the data fetch from [`<ClientContext />`](https://github.com/zetkin/app.zetkin.org/blob/issue-1620/next-v13-survey-page-with-app-router-and-server-action/src/core/env/ClientContext.tsx), where it's hardcoded as `BrowserApiClient`. I think playwright noticed the error when the server render attempted to use the client-side API client and failed the test for it or something like that. Whatever the specifics, fetching this in the server render fixes the test and was what I originally intended to do anyway.

3. ❌ `submits user signature`
Not possible to make this pass until the authentication code is migrated to the app router. Have left the test in with `.skip()` on it because it's not a helpful reason for the build to fail.

4. ❌ `submits anonymous signature`
Actually not sure why this one is failing. Suspect there's something subtly different about the server action requests & responses. Next on the list. Skipping for now for no good reason really, just want the peace of mind of a nice green build.

5. ❌ `preserves inputs on error`
The final boss. Skipping for the same reason as 4.

henrycatalinismith commented:
[92e0d10](https://github.com/zetkin/app.zetkin.org/pull/1756/commits/92e0d105e424952793d4bad3e08cfc71b10427a8) gets the `submits anonymous signature` test enabled and passing.

My first iteration of the app router thing here was actually mega sloppy about the server component / client component boundary. Now there's a clear & coherent boundary where the `<Page />` components in the `src/app` directory are server components that handle data fetching server side and then hand off to client components from the `src/features/survey` directory.

Tidying up that boundary line was enough to get this test passing.

henrycatalinismith commented:
6964e234581b02c21bfa9ce87b9f840a69eba96d has done two things: gotten the `preserves inputs on error` playwright test passing, and broken a bunch of others. The key changes here that made the test pass are as follows.

* Tweak the server action in [`features/surveys/actions/submit.ts`](https://github.com/zetkin/app.zetkin.org/blob/6964e234581b02c21bfa9ce87b9f840a69eba96d/src/features/surveys/actions/submit.ts) so that it returns a value denoting the outcome of the submit attempt.
* Remove the submitted page and instead render the success message on the same URL as the form when the submit attempt's outcome is `success`. I tried to avoid this because I prefer it when people can refresh the success page. But the redirect necessary to enable it doesn't seem to fit within the new server actions philosophy.
* Install canary versions of `react` and `react-dom` in order to be able to use [`useFormState`](https://react.dev/reference/react-dom/hooks/useFormState). Seems to be simply impossible to do error handling under this new paradigm without this hook, and it's not shipped yet.
* Update to version 14 of Next.js. Didn't love that this was a requirement, but it just doesn't work without it. You get a `Converting circular structure to JSON` error during `yarn tsc --noEmit` without it. Seems like some aspect of the canary versions of `react` and `react-dom` upsets TypeScript otherwise.

Bit of a two steps forward one step back commit given the number of new playwright test failures it introduces and the worrying dependency on canary versions of key dependencies. But! BUT! We now finally have a decently clear view here of how it'll probably look to write this kind of code at Zetkin a little while in the future once all this bleeding edge stuff goes mainstream. And god help me I think I actually like it!

richardolsson commented:
I'm just having a look at the [preview deployment of this](https://app-zetkin-oe9b8zsnd-zetkin.vercel.app/o/1/surveys/1), and I'm missing the "What is this question?" question that should have two options. Are multi-choice questions not implemented? Or maybe it's because there is no `response_config` (should default to radio buttons)?

## What's being rendered on the preview
![image](https://github.com/zetkin/app.zetkin.org/assets/550212/0556c06d-c17e-4c92-975a-1730208103cd)

## The API response for this survey
```json
{
   "data" : {
      "access" : "open",
      "allow_anonymous" : true,
      "callers_only" : false,
      "campaign" : null,
      "elements" : [
         {
            "hidden" : false,
            "id" : 1,
            "question" : {
               "description" : "Answer whatever",
               "options" : [
                  {
                     "id" : 1,
                     "text" : "Option 1"
                  },
                  {
                     "id" : 2,
                     "text" : "Option 2"
                  }
               ],
               "question" : "What is this question?",
               "required" : false,
               "response_config" : {},
               "response_type" : "options"
            },
            "type" : "question"
         },
         {
            "hidden" : false,
            "id" : 2,
            "text_block" : {
               "content" : "With some instructions",
               "header" : "This is a text block"
            },
            "type" : "text"
         },
         {
            "hidden" : false,
            "id" : 3,
            "question" : {
               "description" : null,
               "question" : "Is this a free text question?",
               "required" : false,
               "response_config" : {
                  "multiline" : true
               },
               "response_type" : "text"
            },
            "type" : "question"
         }
      ],
      "expires" : null,
      "id" : 1,
      "info_text" : "This is a survey",
      "org_access" : "sameorg",
      "organization" : {
         "id" : 1,
         "title" : "My Organization"
      },
      "published" : "2024-02-22T06:16:19.016317",
      "signature" : "allow_anonymous",
      "title" : "A very open survey"
   }
}
```

henrycatalinismith commented:
Good spot, I’ll get that sorted this evening. Will also have another look at the playwright thing. The canary version of react-dom is what’s behind all the failing tests so I’m keeping an eye open for new versions in case one fixes that.

henrycatalinismith commented:
Right then, it's Code Camp, and I'm kicking off Saturday by rebasing https://github.com/zetkin/app.zetkin.org/pull/1807 into this branch to get the build green. Then the styling work and whatever else emerges along the way to get this thing into a production-ready state. If all goes well this can be a mergeable PR by Sunday afternoon.

henrycatalinismith commented:
So here's what 9f2480dfc39e7ea7c865c892d7c27d95fdd30832 does. The UI before this commit was still in a raw unstyled state, so it was worth doing a quick initial push forwards to get the broad strokes of the mobile layout built. Means I can jump back into detail mode for the rest of the weekend to knock out any mistakes, tweaks, and missing bits like e.g. the desktop layout.

| Before | After |
|-|-|
| ![Screenshot 2024-03-16 at 11-01-14 Next js](https://github.com/zetkin/app.zetkin.org/assets/566159/28b4f1a3-ac1f-42f7-bdb5-6b310fa0c28e) | ![Screenshot 2024-03-16 at 10-55-29 Next js](https://github.com/zetkin/app.zetkin.org/assets/566159/2369815a-5e6c-456e-95fb-2b08145ab4ea) |

henrycatalinismith commented:
Cool cool, got a nice rudimentary desktop layout into [d5741b7](https://github.com/zetkin/app.zetkin.org/pull/1756/commits/d5741b7da1a965b74766e6e7ca14b91fa524a875).

| Before | After |
|-|-|
| ![Screenshot 2024-03-16 at 11-16-42 Next js](https://github.com/zetkin/app.zetkin.org/assets/566159/19c66d69-d76f-4a45-90b6-b18024f0f511) |  ![Screenshot 2024-03-16 at 14-13-54 Next js](https://github.com/zetkin/app.zetkin.org/assets/566159/95860ee5-21ed-405b-a443-21e18a7423fd) |

henrycatalinismith commented:
Another quick win in 62b1adc8360d8960ca3d2cfc96679ba5f8d4bfe3.

| Before | After |
|-|-|
| ![Screenshot 2024-03-16 at 14-26-29 Next js](https://github.com/zetkin/app.zetkin.org/assets/566159/ebee0091-61e1-4ecb-8cca-1a79b7f1098b) | ![Screenshot 2024-03-16 at 14-25-38 Next js](https://github.com/zetkin/app.zetkin.org/assets/566159/6535a034-9e2a-400f-8265-02d0202832e2) |

henrycatalinismith commented:
Big boatload of polish in 7e6263194f46c20539fb0b81e87b066970783525.

| Before | After |
|-|-|
|  ![Screenshot 2024-03-16 at 15-28-35 Next js](https://github.com/zetkin/app.zetkin.org/assets/566159/97f7364c-a8f5-414b-af2b-84040908ff1e) | ![Screenshot 2024-03-16 at 15-27-36 Next js](https://github.com/zetkin/app.zetkin.org/assets/566159/cfdcea74-9857-428b-88ad-9d33f6d66854) |

henrycatalinismith commented:
Spent some time poking around the Vänsterpartiet Malmö member survey and noticed some of the options have quite long text values. So I put together c00f14ea4fb6fcf52c3494fb0c01da8d21b15d6a to ensure the new survey copes with extra long options without the layout collapsing. It's tricky. You can't necessarily make it look _good_ when the text is this long, but you can stop it from breaking. These new pill buttons are a visual improvement on the whole but they do look worse than before for this unusual case. We can probably tweak the border radius here for this case. I spent a bit of time with that but decided to leave it for now and just continue progressing forward.

| Before | After |
|-|-|
| ![Screenshot 2024-03-16 at 16 38 18](https://github.com/zetkin/app.zetkin.org/assets/566159/f5e001d8-9547-403e-bf27-48b09f664ce6) | ![Screenshot 2024-03-16 at 16 38 49](https://github.com/zetkin/app.zetkin.org/assets/566159/8297f51c-bf68-4531-a10b-41024ecbecb9) |

henrycatalinismith commented:
2f3d49c38 adds a `hideOrganization` URL parameter so that it's possible to remove the org banner from the top of the survey when iframing the survey.

```html
<iframe src="http://localhost:3000/o/1/surveys/5?hideOrganization=true"></iframe>
```

| Before | After |
|-|-|
| ![Screen Shot 2024-03-16 at 16 59 04](https://github.com/zetkin/app.zetkin.org/assets/566159/3a9b9c16-200c-42d6-bce5-89fab5442a8c) |  ![Screen Shot 2024-03-16 at 17 01 09](https://github.com/zetkin/app.zetkin.org/assets/566159/33139061-6bf9-4ff8-9312-ef0086a183f7) |

henrycatalinismith commented:
Had to make the dropdown menus controlled in [509aea0](https://github.com/zetkin/app.zetkin.org/pull/1756/commits/509aea000fd79430f9d7332d21a6afc67f3e32cb) in order to fix a warning from MUI about `undefined` being an invalid default value.

| Before | After |
|-|-|
| ![Screenshot 2024-03-16 at 17 29 05](https://github.com/zetkin/app.zetkin.org/assets/566159/4aea646e-3aa4-43a3-9f5c-8e0a6f98266f) | ![Screenshot 2024-03-16 at 17 28 47](https://github.com/zetkin/app.zetkin.org/assets/566159/154640f9-4172-485f-9076-81d1c9056505) |

henrycatalinismith commented:
[682d865](https://github.com/zetkin/app.zetkin.org/pull/1756/commits/682d865c8d72b6bc3c7d3c1503a937df408d8fd5) is based on https://nextjs.org/docs/app/api-reference/functions/generate-metadata

| Before | After |
|-|-|
| ![Screenshot 2024-03-16 at 17 38 28](https://github.com/zetkin/app.zetkin.org/assets/566159/fbcec0b4-4397-4ec6-86fd-46e45c5af779) | ![Screenshot 2024-03-16 at 17 38 10](https://github.com/zetkin/app.zetkin.org/assets/566159/52cb4c03-3031-406f-b63b-370ae84344e0) |

henrycatalinismith commented:
And finally here's aa7c27390fb0bbc3bfea657197df455be3a30fc8 with the URL update.

| Before | After |
|-|-|
| ![Screenshot 2024-03-16 at 18 03 51](https://github.com/zetkin/app.zetkin.org/assets/566159/128c0d97-4a37-4f39-a29b-1799c9167605) |  ![Screenshot 2024-03-16 at 18 03 19](https://github.com/zetkin/app.zetkin.org/assets/566159/8226b29f-4aa1-4f78-a714-19be0b6e1bff) |

henrycatalinismith commented:
My commit c3058429f15c588d137947f3cfb4303d12773dd2 in this branch from back in January contains a mistake. I attempted to use a negative lookahead assertion – `(!.+\\/surveys)` – to exclude the survey page URL from the redirect rule. This didn't work, and instead excluded _all_ the URLs from it. I thought I'd already pushed a commit fixing this, but on closer inspection today I realised I'd only done the thinking part of this and hadn't actually typed it in and pushed it yet! So there it is now anyway in 7a98c28d48e30a8f28d0051c9cb72fa12035ff30!

richardolsson commented on src/app/o/[orgId]/surveys/[surveyId]/page.tsx:
```diff
@@ -0,0 +1,45 @@
+  };
+};
+
+export async function generateMetadata({
```
Beautiful!

richardolsson commented on src/app/o/[orgId]/surveys/[surveyId]/page.tsx:
```diff
@@ -0,0 +1,45 @@
+  };
+}
+
+// @ts-expect-error Async support missing
```
What is the purpose of this comment?

henrycatalinismith replied:
Replaced `Async support missing` with https://nextjs.org/docs/app/building-your-application/configuring/typescript#async-server-component-typescript-error so that the full context is immediately available one click away. We're just a TypeScript major version upgrade away from being able to remove this.

richardolsson commented on integrationTesting/tests/organize/surveys/submitting-survey.spec.ts:
```diff
@@ -0,0 +1,191 @@
+
+    await page.click('input[name="1.options"][value="1"]');
+    await page.fill('[name="2.text"]', 'Topple capitalism');
+    await page.click('input[name="sig"][value="user"]');
```
I see this working in the test (which is passing) but for some reason I'm not seeing this input when I'm using the page while authenticated. Lets have a look together today!

henrycatalinismith replied:
This test is actually skipped!

```
test.skip('submits user signature', async ({ moxy, page }) => {
```

richardolsson commented on src/core/env/ClientContext.tsx:
```diff
@@ -0,0 +1,52 @@
+  interface DefaultTheme extends Theme {}
+}
+
+export type ClientContextProps = {
```
Does this need to be exported?

henrycatalinismith replied:
Removed! I have a defensive coding habit of always exporting prop typedefs in case anything suddenly wants to import them. Doesn't seem to be the normal thing here!

richardolsson commented on src/features/surveys/actions/submit.ts:
```diff
@@ -0,0 +1,23 @@
+): Promise<ZetkinSurveyFormStatus> {
+  const data = Object.fromEntries([...formData.entries()]);
+  const submission = prepareSurveyApiSubmission(data, false);
+  const apiClient = new BackendApiClient({});
```
Why are headers not being forwarded to the `BackendApiClient()` constructor? Unless I'm mistaken, this means that the user will not be authenticated when making this request, which means the "user" signature type can never work.

henrycatalinismith replied:
Fixed!

richardolsson commented on src/features/surveys/components/SurveyURLCard.tsx:
```diff
@@ -37,9 +37,9 @@ const SurveyURLCard = ({ isOpen, orgId, surveyId }: SurveyURLCardProps) => {
       <Box display="flex" paddingBottom={2}>
         <ZUITextfieldToClipboard
-          copyText={`${process.env.NEXT_PUBLIC_ZETKIN_APP_DOMAIN}/o/${orgId}/surveys/${surveyId}`}
+          copyText={`${location.protocol}//${location.host}/o/${orgId}/surveys/${surveyId}`}
```
Perfect, so simple!

richardolsson commented on src/features/surveys/components/surveyForm/SurveySignature.tsx:
```diff
@@ -0,0 +1,135 @@
+            </FormLabel>
+
+            <Box display="flex" flexDirection="column" rowGap={1}>
+              {/* <SurveyOption
+            control={<Radio required />}
+            label={
+              <Typography>
+                <Msg
+                  id={messageIds.surveySignature.type.user}
+                  values={{
+                    email: currentUser?.email ?? '',
+                    person: currentUser?.first_name ?? '',
+                  }}
+                />
+              </Typography>
+            }
+            value="user"
+          /> */}
```
Can we re-enable this (only when user is authenticated)?

richardolsson commented on src/features/surveys/l10n/messageIds.ts:
```diff
@@ -159,14 +158,56 @@ export default makeMessages('feat.surveys', {
   unlinkedCard: {
     description: m(
-      'When someone submits a survey without logging in,that survey will be unlinked. Searching for people in Zetkin based on their survey responses will not work on unlinked submissions.'
+      'When someone submits a survey without logging in, that survey will be unlinked. Searching for people in Zetkin based on their survey responses will not work on unlinked submissions.'
```
Well spotted!

richardolsson commented on src/features/surveys/utils/prepareSurveyApiSubmission.spec.ts:
```diff
@@ -0,0 +1,61 @@
+import prepareSurveyApiSubmission from './prepareSurveyApiSubmission';
+
+describe('prepareSurveyApiSubmission()', () => {
```
Great tests!

richardolsson commented on src/locale/sv.yml:
```diff
@@ -1416,6 +1416,18 @@ feat:
       personRecordColumn: Svarande
       suggestedPeople: Föreslagna personer
       unlink: Avlänka
+    surveyForm:
```
We generally don't change the YAML files manually. Translators provide them through https://translate.zetkin.org. But the fact that they're here made me want to test them, but when I set my browser language to Swedish the UI is still showing the English strings. Any ideas why?

henrycatalinismith replied:
Untranslated new strings!

> Error: [@formatjs/intl Error MISSING_TRANSLATION] Missing message: "feat.surveys.surveyForm.submit" for locale "sv", using default message (Submit) as fallback.

richardolsson commented on src/features/surveys/components/surveyForm/SurveyForm.tsx:
```diff
@@ -0,0 +1,60 @@
+} from 'utils/types/zetkin';
+
+// eslint-disable-next-line @typescript-eslint/ban-ts-comment
+// @ts-ignore
```
What's the problem with this import that requires `ts-ignore`?

henrycatalinismith replied:
Added an explanatory comment for future readers. Ideally the React team would publish typedefs directly within these libs. The DefinitelyTyped definitions for react-dom are lagging behind these new canary version features.

richardolsson commented on src/pages/o/[orgId]/map/index.tsx:
```diff
@@ -0,0 +1,27 @@
+import { FC } from 'react';
+import { scaffold } from 'utils/next';
+
+const scaffoldOptions = {
+  allowNonOfficials: true,
+  authLevelRequired: 1,
+};
+
+export const getServerSideProps = scaffold(async (ctx) => {
+  const { orgId } = ctx.params!;
+
+  return {
+    props: {
+      orgId,
+    },
+  };
+}, scaffoldOptions);
+
+type PageProps = {
+  orgId: string;
+};
+
+const Page: FC<PageProps> = ({ orgId }) => {
+  return <h1>Map page for org {orgId}</h1>;
+};
+
+export default Page;
```
Let's get rid of this!

richardolsson commented on src/pages/o/[orgId]/projects/[campId]/index.tsx:
```diff
@@ -0,0 +1,33 @@
+import { FC } from 'react';
+import { scaffold } from 'utils/next';
+
+const scaffoldOptions = {
+  allowNonOfficials: true,
+  authLevelRequired: 1,
+};
+
+export const getServerSideProps = scaffold(async (ctx) => {
+  const { campId, orgId } = ctx.params!;
+
+  return {
+    props: {
+      campId,
+      orgId,
+    },
+  };
+}, scaffoldOptions);
+
+type PageProps = {
+  campId: string;
+  orgId: string;
+};
+
+const Page: FC<PageProps> = ({ orgId, campId }) => {
+  return (
+    <h1>
+      Page for org {orgId}, project {campId}
+    </h1>
+  );
+};
+
+export default Page;
```
Let's get rid of this!

richardolsson commented on src/pages/o/[orgId]/index.tsx:
```diff
@@ -0,0 +1,27 @@
+import { FC } from 'react';
+import { scaffold } from 'utils/next';
+
+const scaffoldOptions = {
+  allowNonOfficials: true,
+  authLevelRequired: 1,
+};
+
+export const getServerSideProps = scaffold(async (ctx) => {
+  const { orgId } = ctx.params!;
+
+  return {
+    props: {
+      orgId,
+    },
+  };
+}, scaffoldOptions);
+
+type PageProps = {
+  orgId: string;
+};
+
+const Page: FC<PageProps> = ({ orgId }) => {
+  return <h1>Home page for org {orgId}</h1>;
+};
+
+export default Page;
```
Let's get rid of this!

richardolsson commented on src/pages/o/[orgId]/events/[eventId]/index.tsx:
```diff
@@ -0,0 +1,33 @@
+import { FC } from 'react';
+import { scaffold } from 'utils/next';
+
+const scaffoldOptions = {
+  allowNonOfficials: true,
+  authLevelRequired: 1,
+};
+
+export const getServerSideProps = scaffold(async (ctx) => {
+  const { eventId, orgId } = ctx.params!;
+
+  return {
+    props: {
+      eventId,
+      orgId,
+    },
+  };
+}, scaffoldOptions);
+
+type PageProps = {
+  eventId: string;
+  orgId: string;
+};
+
+const Page: FC<PageProps> = ({ orgId, eventId }) => {
+  return (
+    <h1>
+      Page for org {orgId}, event {eventId}
+    </h1>
+  );
+};
+
+export default Page;
```
Let's get rid of this!

richardolsson commented on src/pages/my/todo/index.tsx:
```diff
@@ -0,0 +1,21 @@
+import { FC } from 'react';
+import { scaffold } from 'utils/next';
+
+const scaffoldOptions = {
+  allowNonOfficials: true,
+  authLevelRequired: 1,
+};
+
+export const getServerSideProps = scaffold(async () => {
+  return {
+    props: {},
+  };
+}, scaffoldOptions);
+
+type PageProps = void;
+
+const Page: FC<PageProps> = () => {
+  return <h1>Todo page for currently signed in user</h1>;
+};
+
+export default Page;
```
Let's get rid of this!

richardolsson commented on src/pages/my/home/index.tsx:
```diff
@@ -0,0 +1,21 @@
+import { FC } from 'react';
+import { scaffold } from 'utils/next';
+
+const scaffoldOptions = {
+  allowNonOfficials: true,
+  authLevelRequired: 1,
+};
+
+export const getServerSideProps = scaffold(async () => {
+  return {
+    props: {},
+  };
+}, scaffoldOptions);
+
+type PageProps = void;
+
+const Page: FC<PageProps> = () => {
+  return <h1>Home page for currently signed in user</h1>;
+};
+
+export default Page;
```
Let's get rid of this!

richardolsson reviewed (changes requested):
Awesome work on all of this! There so much amazing stuff going on here. I've played around on the preview build, and I've looked through all of the code, and I did find some things that I think we should change. Let's go over it together if you want to! Comments below include big and small, cheers and requests for changes, and some notes to self for our conversation.

henrycatalinismith commented on src/app/layout.tsx:
```diff
@@ -0,0 +1,26 @@
+import { getBrowserLanguage, getMessages } from 'utils/locale';
+
+export const metadata = {
+  description: 'Generated by Next.js',
+  title: 'Next.js',
```
Removed! Individual pages should always override it. There's currently just the one page in the app router so it's kind of moot for now and probably better just absent entirely.

henrycatalinismith commented on next.config.js:
```diff
@@ -57,10 +57,24 @@ module.exports = {
+        source: '/o/:orgId/map',
+        destination: `http://${process.env.ZETKIN_API_DOMAIN}/o/:orgId/map`,
+        permanent: false,
+      },
```
I just added one of these each for all the placeholder pages from the previous Code Camp. Wasn't sure which ones were needed to avoid breaking existing stuff. Have removed the `/o/:orgId/map` one now as it's clear that one's not needed. LMK if there's others worth removing too.

henrycatalinismith commented:
Found myself a moment to test a full survey containing all the possible options. Here's a screen recording of filling out an example submission.

https://github.com/zetkin/app.zetkin.org/assets/566159/22936e43-247a-4608-b37e-58d6c74447a7

Pay particular attention to the checkbox input values I gave at 00:20. I selected both.

<img width="222" src="https://github.com/zetkin/app.zetkin.org/assets/566159/acaa58c3-486a-4cb4-bce6-f96da807c896" />

Now here's me browsing the result. Look how many checkbox values are listed: only one!

https://github.com/zetkin/app.zetkin.org/assets/566159/130dcc72-acf0-4aed-8a2e-a17f1b3aa0e3

So yeah there's a data-loss bug here! Here's the full JSON dump of the submission data that's stored in the database.

```json
{
	"data": {
		"id": 51,
		"respondent": {
			"id": 2,
			"email": "testadmin@example.com",
			"first_name": "Angela",
			"last_name": "Davis"
		},
		"organization": {
			"id": 1,
			"title": "My Organization"
		},
		"survey": {
			"id": 7,
			"title": "Kitchen Sink Survey"
		},
		"submitted": "2024-03-29T15:09:00.530122+00:00",
		"responses": [
			{
				"question_id": 22,
				"response": "This is the answer to the first question."
			},
			{
				"question_id": 23,
				"response": "And here's an answer to question two.\r\n\r\nIt's very long."
			},
			{
				"question_id": 24,
				"options": [
					179
				]
			},
			{
				"question_id": 25,
				"options": [
					181
				]
			},
			{
				"question_id": 26,
				"options": [
					182
				]
			}
		]
	}
}
```

henrycatalinismith commented:
Fixed in 3bc0f5a3de37eb5f1e443fbeaeb26534aa08fa95.

https://github.com/zetkin/app.zetkin.org/assets/566159/67c051f6-18c4-4a6a-bf5a-65ccc47724fa

It was a mistake not to pass the original `FormData` object through to `prepareSurveyApiSubmission()`, I think. That `Object.fromEntries([...formData.entries()])` conversion we were doing on the data before passing it to the function was where the data loss occurred. If you do e.g. `Object.fromEntries([["a", 1], ["a", 2]])`, you get `{ a: 2 }` as the outcome. So formdata is too funky to distill neatly down to a plain JS object, cos checkbox values are represented as duplicate keys with different values.

My fix in 3bc0f5a3de37eb5f1e443fbeaeb26534aa08fa95 passes the actual `FormData` object instead. This enables the function to use the built-in [`getAll()`](https://developer.mozilla.org/en-US/docs/Web/API/FormData/getAll) method to handle multiple values with the same key.

richardolsson commented:
Great work finding and fixing this @henrycatalinismith! It would be great to have playwright tests testing this too.

henrycatalinismith commented:
Done in dfc47a6bc85e4d476de4563a1a745d3e3b463901!

I also created a separate throwaway branch without the fix from 3bc0f5a3de37eb5f1e443fbeaeb26534aa08fa95 and then tried these new playwright assertions there just for maximum certainty. Here's how they fail when the bug is actually present.

![Screenshot 2024-03-29 at 17 23 02](https://github.com/zetkin/app.zetkin.org/assets/566159/ea0bfb8f-90f2-45d7-b089-548b23c710c7)

henrycatalinismith commented:
Quick side-by-side visual comparison of the old survey vs the new one.

| Old | New |
|-|-|
| ![Screenshot 2024-03-30 at 11-16-09 Vänsterpartiet Malmö Zetkin](https://github.com/zetkin/app.zetkin.org/assets/566159/f208b906-1565-466d-a50a-018bd94d13e4) |  ![Screenshot 2024-03-30 at 11-15-59 Members' Survey 2022 - Vänsterpartiet Malmö](https://github.com/zetkin/app.zetkin.org/assets/566159/15b019b4-2c24-442a-81bf-c906b2c3f343) |

henrycatalinismith commented:
Here's a full completion of the members' survey followed by a visual review of the submission in the organizer UI. Did this while halfway out the door with the kids so may have missed something, but I thought it seemed 100% okay.

https://github.com/zetkin/app.zetkin.org/assets/566159/e9055d25-c335-4cb4-b997-95f57cf066ad

richardolsson commented:
Awesome @henrycatalinismith! Is there currently anything testing the `hidden: false` flag?

henrycatalinismith commented:
Fixed one of these just now. Presumably related to the TypeScript upgrade!

![Screenshot 2024-04-02 at 17 20 56](https://github.com/zetkin/app.zetkin.org/assets/566159/1ea10fea-5ed0-41ff-a337-b94a96323728)

richardolsson commented on src/features/surveys/components/SurveyURLCard.tsx:
```diff
@@ -37,9 +37,9 @@ const SurveyURLCard = ({ isOpen, orgId, surveyId }: SurveyURLCardProps) => {
         <ZUITextfieldToClipboard
-          copyText={`${process.env.NEXT_PUBLIC_ZETKIN_APP_DOMAIN}/o/${orgId}/surveys/${surveyId}`}
+          copyText={`${location.protocol}//${location.host}/o/${orgId}/surveys/${surveyId}`}
         >
-          {`${process.env.NEXT_PUBLIC_ZETKIN_APP_DOMAIN}/o/${orgId}/surveys/${surveyId}`}
+          {`${location.protocol}//${location.host}/o/${orgId}/surveys/${surveyId}`}
```
The `orgId` in both of these places should be `survey.organization.id` (which may or may not be the organization that we're currently working within). You should be able to get it using `useSurvey()`.

See #1941 for related changes (but we refrained from making changes to the URL there to avoid merge conflicts).

richardolsson commented on src/features/surveys/components/surveyForm/SurveySignature.tsx:
```diff
@@ -0,0 +1,139 @@
+};
+
+const SurveySignature: FC<SurveySignatureProps> = ({ survey, user }) => {
+  // const currentUser = useCurrentUser();
```
Why is this currently commented out? It looks to me like the user is coming in through props, so I'm assuming this line could be deleted?

richardolsson commented on src/features/surveys/components/surveyForm/SurveySignature.tsx:
```diff
@@ -0,0 +1,139 @@
+import SurveyContainer from './SurveyContainer';
+import SurveyOption from './SurveyOption';
+import SurveySubheading from './SurveySubheading';
+// import useCurrentUser from 'features/user/hooks/useCurrentUser';
```
This should be deleted if not used

richardolsson reviewed (commented):
I think it's becoming time to merge this. I have some minor questions below, and I think it would be nice to merge main one last time after EOP tomorrow. Then let's get it merged fully!

richardolsson commented:
When you merge `main`, you will find that there are lots of linting errors since we introduced a new rule in #2065, but (almost) all of them should be fixable with `yarn eslint --fix`.

henrycatalinismith commented:
The latest commit in this branch now has a broken build. I've fixed the linter error and removed the commented-out code but there's a new problem. The Storybook fixes have introduced a new react-dom version and it's broken the `useFormState` import from the canary version.

### `yarn why react-dom`

<table><thead><tr><th>Before the Storybook fixes</th><th>After the Storybook fixes</th></tr></thead><tbody><tr><td><pre><code>yarn why v1.22.21
warning ../package.json: No license field
[1/4] 🤔  Why do we have the module "react-dom"...?
[2/4] 🚚  Initialising dependency graph...
[3/4] 🔍  Finding dependency...
[4/4] 🚡  Calculating file sizes...
=> Found "react-dom@18.3.0-canary-763612647-20240126"
info Has been hoisted to "react-dom"
info This module exists because it's specified in "dependencies".
info Disk size without dependencies: "13.36MB"
info Disk size with unique dependencies: "13.7MB"
info Disk size with transitive dependencies: "13.73MB"
info Number of shared dependencies: 3</code></pre></td><td><pre><code>yarn why v1.22.21
warning ../package.json: No license field
[1/4] 🤔  Why do we have the module "react-dom"...?
[2/4] 🚚  Initialising dependency graph...
[3/4] 🔍  Finding dependency...
[4/4] 🚡  Calculating file sizes...
=> Found "react-dom@18.3.0-canary-763612647-20240126"
info Has been hoisted to "react-dom"
info This module exists because it's specified in "dependencies".
info Disk size without dependencies: "13.36MB"
info Disk size with unique dependencies: "13.7MB"
info Disk size with transitive dependencies: "13.73MB"
info Number of shared dependencies: 3
=> Found "@storybook/addon-docs#react-dom@18.3.1"
info This module exists because "@storybook#addon-essentials#@storybook#addon-docs" depends on it.
info Disk size without dependencies: "4.4MB"
info Disk size with unique dependencies: "4.74MB"
info Disk size with transitive dependencies: "4.77MB"
info Number of shared dependencies: 3</code></pre></td></tr></tbody></table>

It looks like this API has also changed in React since we first worked on this branch, which is fair enough given it's marked as experimental. If you look at the docs for [`useActionState`](https://react.dev/reference/react/useActionState) they say

> In earlier React Canary versions, this API was part of React DOM and called useFormState.

I don't have time to investigate this further right now so will simply post these notes here and leave it at that for the time being.

henrycatalinismith commented:
Confirmed that the import still actually works fine and that the type error is a false alarm. I think my earlier theory about this being related to the duplicate react-dom versions was wrong. This seems like it might be related to the TypeScript upgrade instead. I made the following change.

```diff
 import { Box } from '@mui/material';
 import { FC } from 'react';
+// @ts-expect-error Erroneous `Module '"react-dom"' has no exported member 'useFormState'`
 import { useFormState } from 'react-dom';
```

Then tested submitting a survey, and it worked fine. The playwright tests all pass. I'm going to push this so that we have a green build here. From there we can do whatever.

henrycatalinismith commented on src/app/o/[orgId]/surveys/[surveyId]/page.tsx:
```diff
@@ -0,0 +1,57 @@
+  };
+}
+
+// @ts-expect-error https://nextjs.org/docs/app/building-your-application/configuring/typescript#async-server-component-typescript-error
```
Here's the error when removing this comment.

```
Type '({ params }: PageProps) => Promise<ReactElement<any, string | JSXElementConstructor<any>>>' is not assignable to type 'FC<PageProps>'.
  Type 'Promise<ReactElement<any, string | JSXElementConstructor<any>>>' is missing the following properties from type 'ReactElement<any, any>': type, props, keyts(2322)
```

I put a little time into scouting ahead for the React 19 migration in this repo over the weekend. This error reminds me of some of the trouble I found there. Really hoping things become a bit more stable once this big architectural transition by React is in the rear view mirror.

henrycatalinismith commented:
Finally cracked the mystery of the TypeScript error! It's a known [eslint-plugin-import bug where it moves imports but leaves comments behind](https://github.com/import-js/eslint-plugin-import/issues/1450). Zooming out the diff in the screenshot below, you can see the removal in red at the bottom, the addition in green at the top, and the comment left behind in the old location at the bottom.
 
![Screenshot 2024-07-01 at 22 09 56](https://github.com/zetkin/app.zetkin.org/assets/566159/8e857ac3-4692-4a29-b1be-7e31edf7091e)

Nice to have this one tidied up. At this point I'm happy this branch is cleaned up and ready @richardolsson.

henrycatalinismith commented:
omfg 😂
